### PR TITLE
fix: apply changes to the latest version

### PR DESCRIPTION
**Pull request by zcq98**
# Pull Request

- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Bug fixes

c.config.KubeOvnClient.KubeovnV1().Subnets().Update should be applied on the latest subnet, func formatSubnet should return the latest version subnet（same problem to vlan and vpc update）

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 4f0444f</samp>

Refactor and optimize some functions related to subnet, vlan, vpc, and vip initialization and status update in `pkg/controller`. Reduce unnecessary deep copies and controller arguments, and make the functions return updated objects. Use the `Update` method instead of the deprecated `PatchSync` method in the e2e test.

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 4f0444f</samp>

> _Oh we're the coders of the `Controller` struct_
> _And we make it better with every pull request_
> _We refactor and update with skill and with zest_
> _Heave away, me hearties, heave away_

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 4f0444f</samp>

*  Change `formatSubnet`, `calcDualSubnetStatusIP`, and `calcSubnetStatusIP` functions to methods of the `Controller` struct and remove the `c` parameter ([link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-92aa62435bc2b1b2be34f478f650597068bcf35b8d5975d119f91cf78ac54988L104-R104), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-92aa62435bc2b1b2be34f478f650597068bcf35b8d5975d119f91cf78ac54988L158-R158), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-92aa62435bc2b1b2be34f478f650597068bcf35b8d5975d119f91cf78ac54988L615-R617), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L275-R275), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L711-R712), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L718-R721), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L920-R925), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2007-R2012), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2015-R2018), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2037-R2040), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2046-R2049), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2069-R2072), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2084-R2102), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2111-R2114), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2119-R2122), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2127-R2130), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2177-R2180), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L2183-R2189), [link](https://github.com/kubeovn/kube-ovn/pull/3514/files?diff=unified&w=0#diff-c53f89147c39d7d4c3d3046d288a0cea09719776e947bfb3ef01ea29f75b1fb0L485-R487)). This makes the code more consistent and concise, and avoids passing the controller as an argument to every function that needs it.

### CODE REVIEW

File: pkg/controller/init.go [hunk 1]
```diff
@@ -612,9 +612,9 @@ func (c *Controller) initSyncCrdSubnets() error {
 	for _, orisubnet := range subnets {
 		subnet := orisubnet.DeepCopy()
 		if util.CheckProtocol(subnet.Spec.CIDRBlock) == kubeovnv1.ProtocolDual {
-			err = calcDualSubnetStatusIP(subnet, c)
+			_, err = c.calcDualSubnetStatusIP(subnet)
 		} else {
-			err = calcSubnetStatusIP(subnet, c)
+			_, err = c.calcSubnetStatusIP(subnet)
```
Comment (bobz965) [inline pkg/controller/init.go:617]: 后面有接着 update，这个返回值应该赋值，并且 deepcopy 一下

File: pkg/controller/subnet.go [hunk 2]
```diff
@@ -1992,27 +1995,27 @@ func (c *Controller) reconcileU2OInterconnectionIP(subnet *kubeovnv1.Subnet) err
...
 		}
 	}
 	return nil
 }
 
-func calcDualSubnetStatusIP(subnet *kubeovnv1.Subnet, c *Controller) error {
+func (c *Controller) calcDualSubnetStatusIP(subnet *kubeovnv1.Subnet) (*kubeovnv1.Subnet, error) {
 	if err := util.CheckCidrs(subnet.Spec.CIDRBlock); err != nil {
-		return err
+		return subnet, err
```
Comment (bobz965) [inline pkg/controller/subnet.go:2012]: err的时候，感觉返回空比较好，异常的 subnet 应该也没啥用

File: pkg/controller/subnet.go [hunk 3]
```diff
@@ -2081,22 +2084,22 @@ func calcDualSubnetStatusIP(subnet *kubeovnv1.Subnet, c *Controller) error {
...
+	return newSubnet, err
 }
 
-func calcSubnetStatusIP(subnet *kubeovnv1.Subnet, c *Controller) error {
+func (c *Controller) calcSubnetStatusIP(subnet *kubeovnv1.Subnet) (*kubeovnv1.Subnet, error) {
 	_, cidr, err := net.ParseCIDR(subnet.Spec.CIDRBlock)
 	if err != nil {
 		klog.Error(err)
-		return err
+		return subnet, err
```
Comment (bobz965) [inline pkg/controller/subnet.go:2097]: 同上 err 返回空值

File: pkg/controller/vpc.go [hunk 4]
```diff
@@ -876,20 +876,22 @@ func formatVpc(vpc *kubeovnv1.Vpc, c *Controller) error {
...
 		}
 	}
 
 	if changed {
-		if _, err := c.config.KubeOvnClient.KubeovnV1().Vpcs().Update(context.Background(), vpc, metav1.UpdateOptions{}); err != nil {
+		newVpc, err := c.config.KubeOvnClient.KubeovnV1().Vpcs().Update(context.Background(), vpc, metav1.UpdateOptions{})
+		if err != nil {
 			klog.Errorf("failed to update vpc %s: %v", vpc.Name, err)
-			return err
+			return vpc, err
```
Comment (bobz965) [inline pkg/controller/vpc.go:889]: 因为传入的值是vpc.deepcopy, 所以这里的值更新后和formatSubnet保持一致，返回vpc.deepcopy

**Review by bobz965 — CHANGES_REQUESTED** [review]

File: pkg/controller/subnet.go [hunk 5]
```diff
@@ -917,9 +918,11 @@ func (c *Controller) handleUpdateSubnetStatus(key string) error {
 	}
 
 	if util.CheckProtocol(subnet.Spec.CIDRBlock) == kubeovnv1.ProtocolDual {
-		return calcDualSubnetStatusIP(subnet, c)
+		_, err := c.calcDualSubnetStatusIP(subnet)
```
Comment (bobz965) [inline pkg/controller/subnet.go:921]: err 没有打印

File: pkg/controller/subnet.go [hunk 6]
```diff
@@ -708,16 +709,16 @@ func (c *Controller) handleAddOrUpdateSubnet(key string) error {
...
-	subnet, err := formatSubnet(cachedSubnet.DeepCopy(), c)
+	subnet, err := c.formatSubnet(cachedSubnet.DeepCopy())
 	if err != nil {
 		klog.Error(err)
 		return err
 	}
 
 	if subnet.Spec.Protocol == kubeovnv1.ProtocolDual {
-		err = calcDualSubnetStatusIP(subnet, c)
+		subnet, err = c.calcDualSubnetStatusIP(subnet)
```
Comment (bobz965) [inline pkg/controller/subnet.go:719]: err 没有打印

File: pkg/controller/subnet.go [hunk 7]
```diff
@@ -917,9 +918,11 @@ func (c *Controller) handleUpdateSubnetStatus(key string) error {
 	}
 
 	if util.CheckProtocol(subnet.Spec.CIDRBlock) == kubeovnv1.ProtocolDual {
-		return calcDualSubnetStatusIP(subnet, c)
+		_, err := c.calcDualSubnetStatusIP(subnet)
+		return err
 	}
-	return calcSubnetStatusIP(subnet, c)
+	_, err = c.calcSubnetStatusIP(subnet)
```
Comment (bobz965) [inline pkg/controller/subnet.go:924]: err 没有打印

File: pkg/controller/subnet.go [hunk 8]
```diff
@@ -1992,27 +1995,27 @@ func (c *Controller) reconcileU2OInterconnectionIP(subnet *kubeovnv1.Subnet) err
 		klog.Infof("reconcile underlay subnet %s  to overlay interconnection with U2OInterconnection %v U2OInterconnectionIP %s ",
 			subnet.Name, subnet.Spec.U2OInterconnection, subnet.Status.U2OInterconnectionIP)
 		if subnet.Spec.Protocol == kubeovnv1.ProtocolDual {
-			if err := calcDualSubnetStatusIP(subnet, c); err != nil {
+			if _, err := c.calcDualSubnetStatusIP(subnet); err != nil {
```
Comment (bobz965) [inline pkg/controller/subnet.go:1998]: err 没有打印

File: pkg/controller/subnet.go [hunk 9]
```diff
@@ -1992,27 +1995,27 @@ func (c *Controller) reconcileU2OInterconnectionIP(subnet *kubeovnv1.Subnet) err
...
 				return err
 			}
 		} else {
-			if err := calcSubnetStatusIP(subnet, c); err != nil {
+			if _, err := c.calcSubnetStatusIP(subnet); err != nil {
```
Comment (bobz965) [inline pkg/controller/subnet.go:2002]: err 没有打印

File: pkg/controller/vip.go [hunk 10]
```diff
@@ -482,9 +482,9 @@ func (c *Controller) handleUpdateVirtualParents(key string) error {
 func (c *Controller) subnetCountIP(subnet *kubeovnv1.Subnet) error {
 	var err error
 	if util.CheckProtocol(subnet.Spec.CIDRBlock) == kubeovnv1.ProtocolDual {
-		err = calcDualSubnetStatusIP(subnet, c)
+		_, err = c.calcDualSubnetStatusIP(subnet)
```
Comment (bobz965) [inline pkg/controller/vip.go:485]: err 没有打印

**Review by bobz965 — CHANGES_REQUESTED** [review]

**Review by bobz965 — APPROVED** [review]

**Comment by bobz965** [comment]
hi @oilbeater could you help review this? thanks

File: pkg/controller/subnet.go [hunk 11]
```diff
@@ -708,16 +709,17 @@ func (c *Controller) handleAddOrUpdateSubnet(key string) error {
 	}
 	klog.V(3).Infof("handle add or update subnet %s", cachedSubnet.Name)
 
-	subnet, err := formatSubnet(cachedSubnet.DeepCopy(), c)
+	subnet := cachedSubnet.DeepCopy()
```
Comment (oilbeater) [inline pkg/controller/subnet.go:712]: 可以合并成一行

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/controller/subnet.go (hunk 11 above)
Comment (zcq98) [reply pkg/controller/subnet.go:712]: 其它地方也都是先deepcopy再传参，这样写感觉代码更统一一点

**Review by oilbeater — APPROVED** [review]

**Comment by oilbeater** [comment]
Nice catch! @zcq98

**Pull request merged**